File: 프로그래머스/3/42579. 베스트앨범/베스트앨범.cpp
#include <string>
#include <vector>
#include <unordered_map>
#include <algorithm>

using namespace std;

vector<int> solution(vector<string> genres, vector<int> plays) {
    vector<int> answer;

    // 장르별 총 재생 수
    unordered_map<string, int> genreTotal;

    // 장르별 노래 목록: (고유번호, 재생수)
    unordered_map<string, vector<pair<int, int>>> genreSongs;

    for (int i = 0; i < (int)genres.size(); i++) {
        genreTotal[genres[i]] += plays[i];
        genreSongs[genres[i]].push_back({i, plays[i]});
    }

    // 장르를 총 재생 수 기준으로 정렬하기 위해 벡터로 변환
    vector<pair<string, int>> sortedGenres;
    for (auto &g : genreTotal) {
        sortedGenres.push_back({g.first, g.second});
    }

    sort(sortedGenres.begin(), sortedGenres.end(),
         [](const pair<string, int>& a, const pair<string, int>& b) {
             return a.second > b.second; // 총 재생 수 내림차순
         });

    // 각 장르에서 상위 2곡 선택
    for (auto &g : sortedGenres) {
        string genre = g.first;
        auto &songs = genreSongs[genre];

        sort(songs.begin(), songs.end(),
             [](const pair<int, int>& a, const pair<int, int>& b) {
                 if (a.second == b.second) return a.first < b.first; // 재생 수 같으면 고유번호 오름차순
                 return a.second > b.second; // 재생 수 내림차순
             });

        int cnt = min(2, (int)songs.size());
        for (int i = 0; i < cnt; i++) {
            answer.push_back(songs[i].first);
        }
    }

    return answer;
}

File: 프로그래머스/3/42579. 베스트앨범/베스트앨범.kt
class Solution {
    fun solution(genres: Array<String>, plays: IntArray): IntArray {
        // 장르별 총 재생 수
        val genreTotal = mutableMapOf<String, Int>()

        // 장르별 노래 목록: Pair(고유번호, 재생수)
        val genreSongs = mutableMapOf<String, MutableList<Pair<Int, Int>>>()

        for (i in genres.indices) {
            val genre = genres[i]
            val play = plays[i]

            genreTotal[genre] = (genreTotal[genre] ?: 0) + play
            genreSongs.getOrPut(genre) { mutableListOf() }.add(i to play)
        }

        // 장르를 총 재생 수 기준 내림차순 정렬
        val sortedGenres = genreTotal.entries
            .sortedByDescending { it.value }
            .map { it.key }

        val result = mutableListOf<Int>()

        for (genre in sortedGenres) {
            val songs = genreSongs[genre] ?: mutableListOf()

            // 장르 내 정렬
            val sortedSongs = songs.sortedWith(
                compareByDescending<Pair<Int, Int>> { it.second } // 재생수 내림차순
                    .thenBy { it.first } // 같으면 고유번호 오름차순
            )

            // 최대 2곡 추가
            for (i in 0 until minOf(2, sortedSongs.size)) {
                result.add(sortedSongs[i].first)
            }
        }

        return result.toIntArray()
    }
}